cpp/include/cudf/detail/labeling/label_segments.cuh

```diff
@@ -0,0 +1,96 @@
+ * `offsets[i+1] - offsets[i]`.
+ *
+ * The labels always start from `0` regardless of the offset values.
+ * In case there are empty segments, their corresponding label values will be skipped in the output.
+ *
+ * Note that the caller is responsible to make sure the output range have the correct size, which is
+ * the total segment sizes (i.e., `size = *(offsets_end - 1) - *offsets_begin`). Otherwise, the
+ * result is undefined.
+ *
+ * @code{.pseudo}
+ * Examples:
+ *
+ * offsets = [ 0, 4, 6, 6, 6, 10 ]
+ * output  = [ 0, 0, 0, 0, 1, 1, 4, 4, 4, 4 ]
+ *
+ * offsets = [ 5, 10, 12 ]
+ * output  = [ 0, 0, 0, 0, 0, 1, 1 ]
+ * @endcode
+ *
+ * @param offsets_begin The beginning of the offsets that define segments.
+ * @param offsets_end The end of the offsets that define segments.
+ * @param out_begin The beginning of the output label range.
+ * @param out_end The end of the output label range.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ */
+template <typename InputIterator, typename OutputIterator>
+void label_segments(InputIterator offsets_begin,
+                    InputIterator offsets_end,
+                    OutputIterator out_begin,
+                    OutputIterator out_end,
+                    rmm::cuda_stream_view stream)
+{
+  // If the output array is empty, that means we have all empty segments.
+  // In such cases, we must terminate immediately. Otherwise, the for loop below may try to access
+  // memory of the output array, resulting in "illegal memory access" error.
+  if (thrust::distance(out_begin, out_end) == 0) { return; }
+
+  // When the output array is not empty, always fill it with `0` value first.
+  using OutputType = typename thrust::iterator_value<OutputIterator>::type;
+  thrust::uninitialized_fill(rmm::exec_policy(stream), out_begin, out_end, OutputType{0});
+
+  // If the offsets array has no more than 2 offset values, there will be at max 1 segment.
+  // In such cases, the output will just be an array of all `0` values (which we already filled).
+  if (thrust::distance(offsets_begin, offsets_end) <= 2) { return; }
+
+  thrust::for_each(rmm::exec_policy(stream),
+                   offsets_begin + 1,  // exclude the first offset value
+                   offsets_end - 1,    // exclude the last offset value
+                   [offsets = offsets_begin, output = out_begin] __device__(auto const idx) {
+                     // Zero-normalized offsets.
+                     auto const dst_idx = idx - (*offsets);
+
+                     // Scatter value `1` to the index at (idx - offsets[0]).
+                     // In case we have repeated offsets (i.e., we have empty segments), this
+                     // `atomicAdd` call will make sure the label values corresponding to these
+                     // empty segments will be skipped in the output.
+                     atomicAdd(&output[dst_idx], OutputType{1});
+                   });
```

Comment:
This looks strange. Why isn't this just a normal `scatter` using the `offsets` as the scatter map?

Reply:
If using just `scatter`, we will have the `output` (sparsely) containing just `1` values. 
If we have the input offsets like `[0, 1, 1, 1, 1, 1, 2]`, `scatter` will result in `[0, 1]`, so the final labels will be just `[0, 1]`. We should have `[0, 5]` instead.

Reply:
The comments above should clarify that. If it still can't do its job, please suggest changes :smile:

Reply:
Ah, I see now, it's because of the empty segments. Hm, tricky. Using an atomicAdd for writing every output is far from ideal.